api/jobs/batch.py

```diff
@@ -0,0 +1,182 @@
+      - have no solution to the gear's input schema (not matched)
+      - have multiple solutions to the gear's input schema (ambiguous)
+      - match the gear's input schema 1 to 1 (matched)
+    Containers are placed in one of the three categories in order.
+    A container with 2 possible files for one input and none for the other
+    will be marked as 'not matched', not ambiguous.
+    """
+
+    matched_conts = []
+    not_matched_conts = []
+    ambiguous_conts = []
+
+    for c in containers:
+        files = c.get('files')
+        if files:
+            suggestions = gears.suggest_for_files(gear, files)
+
+            # Determine if any of the inputs are ambiguous or not satisfied
+            ambiguous = False # Are any of the inputs ambiguous?
+            not_matched = False
+            for files in suggestions.itervalues():
+                if len(files) > 1:
+                    ambiguous = True
+                elif len(files) == 0:
+                    not_matched = True
+                    break
+
+            # Based on results, add to proper list
+            if not_matched:
+                not_matched_conts.append(c)
+            elif ambiguous:
+                ambiguous_conts.append(c)
+            else:
+                # Create input map of file refs
+                inputs = {}
+                for input_name, files in suggestions.iteritems():
+                    inputs[input_name] = {'type': container_type, 'id': str(c['_id']), 'name': files[0]}
+                c['inputs'] = inputs
+                matched_conts.append(c)
+        else:
+            not_matched.append(c)
+    return {
+        'matched': matched_conts,
+        'not_matched': not_matched_conts,
+        'ambiguous': ambiguous_conts
+    }
+
+def insert(batch_proposal):
+    """
+    Simple database insert given a batch proposal.
+    """
+    return config.db.batch.insert(batch_proposal)
+
+def update(batch_id, payload):
+    """
+    Updates a batch job, being mindful of state flow.
+    """
+
+    bid = bson.ObjectId(batch_id)
+    query = {'_id': bid}
+    if payload.get('state'):
+        # Require that the batch job has the previous state
+        query['state'] = BATCH_JOB_TRANSITIONS[payload.get('state')]
+    result = config.db.batch.update_one({'_id': bid}, {'$set': payload})
+    if result.modified_count != 1:
+        raise Exception('Batch job not updated')
+
+def run(batch_job):
+    """
+    Creates jobs from proposed inputs, returns jobs enqueued.
+    """
+
+    proposed_inputs = batch_job.get('proposed_inputs', [])
+    gear_name = batch_job.get('gear')
+    config_ = batch_job.get('config')
+    origin = batch_job.get('origin')
+
+    jobs = []
+    job_ids = []
+    for inputs in proposed_inputs:
+        for input_name, fr in inputs.iteritems():
+            inputs[input_name] = create_filereference_from_dictionary(fr)
+        # TODO support analysis gears (will have to create analyses here)
```

Comment:
We'll need to create analyses here on the fly. I've got a plan for this, will add that functionality tomorrow.